AP_Filesystem: LittleFS: clarify fsync bytes logic

tpwrules opened the pull request:
Removes some impossible cases and better documents it.

Tested on KakuteH7Mini-Nand that there are no performance regressions or misaligned fsyncs.

andyp1per commented:
You'll need to walk me through this a bit more to convince me as there are a few things that have changed.

tpwrules commented:
I've split up the commits and put a lot more clarification in the commit messages.

I tested by making sure the same number of blocks per second were read and written, and that there were no misaligned fsyncs (a diagnostic I added a bit ago).

andyp1per commented on libraries/AP_Filesystem/AP_Filesystem_FlashMemory_LittleFS.cpp:
```diff
@@ -421,10 +421,37 @@ uint32_t AP_Filesystem_FlashMemory_LittleFS::bytes_until_fsync(int fd)
+    uint32_t block_index = BLOCK_INDEX(file_pos, block_size);
+    uint32_t block_offset = BLOCK_OFFSET(file_pos, block_size, block_index);
+    if (block_size - block_offset <= nbytes) {
+        if (block_size == block_offset) {
```
So in this path block_size is returned rather than false for no sync, I don't see how the previous PR can have been correct if this is correct.

andyp1per commented on libraries/AP_Filesystem/AP_Filesystem_FlashMemory_LittleFS.cpp:
```diff
@@ -614,34 +641,6 @@ void AP_Filesystem_FlashMemory_LittleFS::mark_dead()
-    if (blocksize - block_offset <= nbytes) {
-        if (blocksize == block_offset) {
-            // exactly at the end of the block, sync and then write all the data
-            AP::FS().fsync(_write_fd);
```
nbytes is unset here but the return value is ignored

tpwrules replied:
That case can never occur, the fsync is never executed.

andyp1per reviewed (changes requested):
Need to understand that the logic change the previous PR made was in fact correct
